### PR TITLE
H3_Oppel_2

### PR DESCRIPTION
**Pull request by philippoppel**
## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: src/plugins/length/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+- infos/licence = BSD
+- infos/needs =
+- infos/provides = check
+- infos/recommends =
+- infos/placements = presetstorage
+- infos/status = maintained unittest nodep
+- infos/metadata = check/length
+- infos/description = validates if input is less or equal to length and throws error otherwise
+
+# IP Address Validation
```
Comment (markus2330) [inline src/plugins/length/README.md:12]: wrong copy&paste

File: src/plugins/length/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+## Installation
+
+See [installation](/doc/INSTALL.md).
+The package is called `libelektra5-experimental`.
+
+## Usage
+
+```sh
+# Mount `length` plugin to cascading namespace `/tests/length`
+kdb mount config.dump /tests/length dump length
```
Comment (markus2330) [inline src/plugins/length/README.md:27]: `TEST_REQUIRED_PLUGINS dump` needed in CMakeLists.txt

File: src/plugins/length/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+- infos = Information about the len plugin is in keys below
+- infos/author = Philipp Oppel <philipp.oppel@tuwien.ac.at>
+- infos/licence = BSD
+- infos/needs =
+- infos/provides = check
+- infos/recommends =
+- infos/placements = presetstorage
+- infos/status = maintained unittest nodep
+- infos/metadata = check/length
```
Comment (markus2330) [inline src/plugins/length/README.md:9]:
```suggestion
- infos/metadata = check/length/max
```

as adding an additional check/length/min will be also useful at some point.

File: src/plugins/length/length.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ * @file
+ *
+ * @brief Source for length plugin
+ *
+ * @copyright BSD License (see doc/LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include <kdberrors.h>
+#include <regex.h>
```
Comment (markus2330) [inline src/plugins/length/length.c:11]: regex is not needed, please remove.

File: src/plugins/length/length.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	else
+	{
+		rc = false;
+	}
+
+
+	if (!rc)
+	{
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (parentKey, "Validation of key %s with value %s failed", keyName (key),
```
Comment (markus2330) [inline src/plugins/length/length.c:45]: Please improve the error message: write that the length check failed and with which length.

File: src/plugins/length/test_length.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#ifdef HAVE_FEATURES_H
+
+#include <features.h>
+// The function `getaddrinfo` used in the `network` plugin leaks memory, if we use the default value for `ai_flags` on systems that use
+// `glibc` 2.19.
+// See also: https://travis-ci.org/ElektraInitiative/libelektra/builds/428298531
+#if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#if !(__GLIBC_PREREQ(2, 20))
+#include <string.h>
+#define PLUGIN_LEAKS_MEMORY (strcmp (PLUGIN_NAME, "network") == 0)
```
Comment (markus2330) [inline src/plugins/length/test_length.h:25]: copy&paste error?

File: src/plugins/length/testmod_length.c [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/**
+ * @file
+ *
+ * @brief Tests for lengthgth plugin
```
Comment (markus2330) [inline src/plugins/length/testmod_length.c:4]: copy&paste error

File: src/plugins/length/testmod_length.c [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @copyright BSD License (see doc/LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include <stdio.h>
+
+#define PLUGIN_NAME "lengthgth"
```
Comment (markus2330) [inline src/plugins/length/testmod_length.c:12]: copy&paste error

**Review by markus2330 — COMMENTED** [review]
Thank you! Please fix build server issues.

**Review by kodebach — CHANGES_REQUESTED** [review]
In principle, this plugin is well written, but there are few major issue with the actual length validation part.
1. `Key`s can be binary. Your code does not take this into account at all. You basically have two options (both of which are acceptable, as long as they are properly documented in the README):
    - Don't support binary keys, and returns a descriptive error if a binary key is marked with `check/length`.
    - Fully support binary keys and validate their length correctly.
2. I don't know how much experience you have with C, but using a function like `sscanf` without proper error handling is not a good idea.
3. You shouldn't manually iterate over a string just to get its length. In general `strlen` is a better idea. It is easier to read and less error prone. Also Elektra already stores the sizes of key values, so no need to recalculate it.

Also, `ksRewind`/`ksNext` are deprecated (even though they aren't marked as such). We will accept the PR even, if you don't replace them, but since you have to change quite a bit of code anyway, it would be nice, if you could use the `for` loop I suggested instead.

File: src/plugins/length/test_length.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
+/**
```
[the remark this answers is not in the document]
Comment (philippoppel) [reply src/plugins/length/test_length.h:1]: I tried to do it as consistent to the other plugins as possible, I looked at ipaddr and did the same. But there is no other reason

**Review by kodebach — APPROVED** [review]

**Comment by kodebach** [comment]
jenkins build libelektra please

**Comment by kodebach** [comment]
@mpranj There seems to be some sort of permission issue on the Jenkins Server. Seems like not even pulling from Github works...

**Comment by kodebach** [comment]
Reopened as #3922

**Pull request closed without merging**